server/src/main/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderator.java

```diff
@@ -0,0 +1,1350 @@
+          throw new ISE("Missing hydrant [%,d] in identifier [%s].", hydrants.size(), identifier);
+        }
+
+        hydrants.add(
+            new FireHydrant(
+                new QueryableIndexSegment(indexIO.loadIndex(hydrantDir), identifier.asSegmentId()),
+                hydrantNumber
+            )
+        );
+      }
+
+      Sink currSink = new Sink(
+          identifier.getInterval(),
+          schema,
+          identifier.getShardSpec(),
+          identifier.getVersion(),
+          tuningConfig.getAppendableIndexSpec(),
+          tuningConfig.getMaxRowsInMemory(),
+          maxBytesTuningConfig,
+          null,
+          hydrants
+      );
+      currSink.finishWriting(); // this sink is not writable
+      //sinks.put(identifier, currSink);
+      return new Pair<>(identifier, currSink);
+    }
+    catch (IOException e) {
+      log.makeAlert(e, "Problem loading sink[%s] from disk.", schema.getDataSource())
+         .addData("identifier path", identifierPath)
+         .emit();
+    }
+    return null;
+  }
+
+  private ListenableFuture<?> removeSink(
+      final SegmentIdWithShardSpec identifier,
+      final Sink sink,
+      final boolean removeOnDiskData
+  )
+  {
+    // Ensure no future writes will be made to this sink.
+    if (sink.finishWriting()) {
+      // Decrement this sink's rows from the counters. we only count active sinks so that we don't double decrement,
+      // i.e. those that haven't been persisted for *InMemory counters, or pushed to deep storage for the total counter.
+      rowsCurrentlyInMemory.addAndGet(-sink.getNumRowsInMemory());
+      bytesCurrentlyInMemory.addAndGet(-sink.getBytesInMemory());
+      bytesCurrentlyInMemory.addAndGet(-calculateSinkMemoryInUsed(sink));
+      for (FireHydrant hydrant : sink) {
+        // Decrement memory used by all Memory Mapped Hydrant
+        if (!hydrant.equals(sink.getCurrHydrant())) {
+          bytesCurrentlyInMemory.addAndGet(-calculateMemoryUsedByHydrants(hydrant));
+        }
+      }
+      // totalRows are not decremented when removing the sink from memory, sink was just persisted and it
+      // still "lives" but it is in hibernation. It will be revived later just before push.
+    }
+
+    // Wait for any outstanding pushes to finish, then abandon the segment inside the persist thread.
+    return Futures.transform(
+        pushBarrier(),
+        new Function<Object, Void>()
+        {
+          @Nullable
+          @Override
+          public Void apply(@Nullable Object input)
+          {
+            if (!sinks.remove(identifier, sink)) {
+              log.error("Sink for segment[%s] no longer valid, not abandoning.", identifier);
+              return null;
+            }
+
+            metrics.setSinkCount(sinks.size());
+
+            for (FireHydrant hydrant : sink) {
+              if (cache != null) {
+                cache.close(SinkQuerySegmentWalker.makeHydrantCacheIdentifier(hydrant));
```

Comment:
Is this needed?

Reply:
not needed, removed along with cache data member